providers/netty4/src/main/java/org/asynchttpclient/netty/request/body/NettyReactiveStreamBody.java

```diff
@@ -0,0 +1,140 @@
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.typesafe.netty.HandlerSubscriber;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.handler.codec.http.DefaultHttpContent;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.concurrent.EventExecutor;
+
+public class NettyReactiveStreamBody implements NettyBody {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyReactiveStreamBody.class);
+    private static final String NAME_IN_CHANNEL_PIPELINE = "request-body-streamer";
+
+    private final Publisher<ByteBuffer> publisher;
+
+    public NettyReactiveStreamBody(Publisher<ByteBuffer> publisher) {
+        this.publisher = publisher;
+    }
+
+    @Override
+    public long getContentLength() {
+        return -1L;
+    }
+
+    @Override
+    public String getContentType() {
+        return null;
+    }
+
+    @Override
+    public void write(Channel channel, NettyResponseFuture<?> future) throws IOException {
+        if (future.isStreamWasAlreadyConsumed()) {
+            LOGGER.warn("Stream has already been consumed and cannot be reset");
+        } else {
+            future.setStreamWasAlreadyConsumed(true);
+            NettySubscriber subscriber = new NettySubscriber(channel);
+            channel.pipeline().addLast(NAME_IN_CHANNEL_PIPELINE, subscriber);
+            publisher.subscribe(new SubscriberAdapter(subscriber));
+        }
+    }
+
+    private static class SubscriberAdapter implements Subscriber<ByteBuffer> {
+        private volatile Subscriber<HttpContent> subscriber;
+        
+        public SubscriberAdapter(Subscriber<HttpContent> subscriber) {
+            this.subscriber = subscriber;
+        }
+        @Override
+        public void onSubscribe(Subscription s) {
+           subscriber.onSubscribe(s);
+        }
+        @Override
+        public void onNext(ByteBuffer t) {
+            ByteBuf buffer = Unpooled.wrappedBuffer(t.array());
+            HttpContent content = new DefaultHttpContent(buffer);
+            subscriber.onNext(content);
+        }
+        @Override
+        public void onError(Throwable t) {
+            subscriber.onError(t);
+        }
+        @Override
+        public void onComplete() {
+            subscriber.onComplete();
+        }        
+    }
+    
+    private static class NettySubscriber extends HandlerSubscriber<HttpContent> {
+        private static final Logger LOGGER = LoggerFactory.getLogger(NettySubscriber.class);
+
+        private final Channel channel;
+
+        public NettySubscriber(Channel channel) {
+            super(channel.eventLoop());
+            this.channel = channel;
+        }
+
+        @Override
+        public void onComplete() {
```

Comment:
`HandlerSubscriber` actually provides `protected complete()` for the purposes of overriding.  If you override that, then lets say in future we find a bug that needs to be fixed by doing some internal processing in `onComplete`, overriding `complete` won't interfere with that.  A good example of this is that currently we let netty buffer all writes, but maybe in future we might find that we need to do our own buffering, in which case, our `onComplete` handling will need to flush before invoking `complete`, but if we change that this will break, since it will bypass the flush.